proposals/BVAL-508.md

```diff
@@ -0,0 +1,257 @@
+    }
+    
+    public interface ManyContainerValuesExtractor<CONTAINER,CONTAINED> {
+        Iterable<CONTAINED> extractValue(CONTAINER container);
+    }
+    
+    // Examples
+    
+    //TODO can T remain unspecified? or should it be ? or Object?
+    public class OptionalContainerExtractor<T> implements SingleContainerValueExtractor<Optional<T>,T> {
+        T extractValue(Optional<T> optional) { ... };
+    }
+
+Bean Validation will have out of the box support for containers for `Iterable` and `Map`.
+
+TODO: any other Java SE type to consider?
+TODO: should we support JavaFX Property<T>?
+
+Custom container extractors can be provided via:
+
+* property
+* factory
+* service locator
+
+TODO: refine how the container implementations are discovered.
+See [Hibernate Validator's feature](http://docs.jboss.org/hibernate/validator/5.2/reference/en-US/html_single/#section-value-handling) 
+
+#### When are containers used?
+
+When a property is of a known container, constraints declarations are explored.
+A typical example is a property using constraints on type parameters of `TYPE_USE` (`Collection<@Size String>`).
+
+If constraints are discovered, the unwrapping mechanism is used
+to read contained value(s) and potentially apply constraints on them (see below).
+
+By constraints, we also mean property traversal `@Valid`.
+
+Containers of containers are applied recursively if necessary.
+TODO: infinite loops?
+
+#### Which container unwrapper to use?
+
+The container used is the one corresponding to the most specific super type
+as defined in the BV resolution of constraint validators.
+
+### Constraints applied and containers
+
+Constraints declared on the type parameter of a type use will be applied to the contained value
+as extracted by the container logic.
+
+    // each String of the collection is validated for the regexp
+    Set<@Pattern(...) String> emails;
+
+By default, constraints declared on the container will apply to the container.
+This ensures backward compatibility.
+
+    // @Size is applied on the collection
+    @Size(min=5) List<Integer> ages;
+
+Extractors can specify that constraints declared on the container apply to the contained value(s);
+
+    @ConstraintsApplyTo(CONTAINED_VALUES)
+    public class JavaFXPropertyContainerExtractor<T> implements SingleContainerValueExtractor<Property<T>,T> { ... }
+    
+    // test that the age is at least 5
+    @Min(5) IntegerProperty<Integer> age;
+
+This is useful for JavaFX to force the validation of the contained properties.
+
+One can also force the constraints to apply to the container or the container value per site
+
+    // the list must have 5 elements at least
+    @ConstraintsApplyTo(CONTAINED_VALUES)
+    @Size(min=5)
+    Optional<List<Integer>> ages;
+    
+    class IntegerList extends ArrayList<Integer> {};
+    
+    // the list must have 5 elements at least
+    @ConstraintsApplyTo(CONTAINED_VALUES)
+    @Min(2)
```

Comment:
@gunnarmorling I fear the rules would be quite complex.
Remember, we are talking about full stack users, not laser focused expert on Bean Validation.
If we can come up with an ultra simple rule, I am happy to reconsider but I have been burned by the groups feature complexity. Not to mention default sequence.